### PR TITLE
Magpie thredds

### PR DESCRIPTION
**Pull request by fmigneault**
- add custom config to allow display of UI catalog elements of thredds service protected behind magpie/twitcher
- bump versions to 3.x to get custom config stored directly in db

### CODE REVIEW

**Comment by MatProv** [comment]
run tests

**Review by MatProv — APPROVED** [review]

File: birdhouse/config/magpie/providers.cfg.template
Comment (tlvu) [inline birdhouse/config/magpie/providers.cfg.template:69]: Sorry, didn't bring this up earlier, did not yet had the chance to have a close look at this config file.

What is `file_patterns` exactly?  Effective for `data_type` only?  How is this different than the pattern (ex: `\\w+\\.ico`) in `metadata_type.prefixes`?

File: birdhouse/config/magpie/providers.cfg.template
Comment (tlvu) [inline birdhouse/config/magpie/providers.cfg.template:79]: I have a vague  memory there are more here?

File: birdhouse/config/magpie/providers.cfg.template
Comment (tlvu) [inline birdhouse/config/magpie/providers.cfg.template:84]: same here, I think we had more.

**Review by tlvu — CHANGES_REQUESTED** [review]
I think there are some more prefixes, will get back to you after our weekly meeting.

File: birdhouse/config/magpie/providers.cfg.template
Comment (fmigneault) [reply birdhouse/config/magpie/providers.cfg.template:69]: description/example of all the parameters here: 
https://pavics-magpie.readthedocs.io/en/3.3.0/services.html#servicethredds

File: birdhouse/config/magpie/providers.cfg.template
Comment (fmigneault) [reply birdhouse/config/magpie/providers.cfg.template:79]: Could be, those are the ones I was asked to put via other comments in original PR that added that feature.

Comment (tlvu) [reply birdhouse/config/magpie/providers.cfg.template:79]: The list was right here https://github.com/Ouranosinc/Magpie/pull/353#issuecomment-710495259.  Missing uddc and iso.

File: birdhouse/config/magpie/providers.cfg.template
Comment (tlvu) [reply birdhouse/config/magpie/providers.cfg.template:84]: Missing wcs and wms.  Ouranos config do not have dap4.

File: birdhouse/config/magpie/providers.cfg.template
Comment (tlvu) [reply birdhouse/config/magpie/providers.cfg.template:69]:
```
# define which pattern matches that will map different path variations into same file resources
# this can be used to consider two file extensions as the same resource to avoid duplication of permissions
file_patterns:
```
Not very clear when reading this explanation.  But since we also have `.ncml` and `.txt` file on Thredds catalog, these should also be added there?

For "this can be used to consider two file extensions as the same resource to avoid duplication of permissions", can you add an example how to configure that to the doc?  I think Ouranos do not have that case yet (2 file extensions as same resource), do you have a live example of this case on some other Thredds server?

In fact, that doc will be very useful to refer back to, can you add a link to that doc directly in the config file so easy reference later?

**Comment by tlvu** [comment]
@fmigneault reading the docs again, it was not clear to me if user omit the `configuration` section, are defaults assigned?  If yes, what are the default values?  Maybe add a link to a sample config file with default values so those default values are easily found.

File: birdhouse/config/magpie/providers.cfg.template
Comment (fmigneault) [reply birdhouse/config/magpie/providers.cfg.template:69]: There is an example in the text below that sample config. 
Will add the link for reference.

**Comment by fmigneault** [comment]
> reading the docs again, it was not clear to me if user omit the `configuration` section, are defaults assigned? If yes, what are the default values? Maybe add a link to a sample config file with default values so those default values are easily found.

Just above the config in https://pavics-magpie.readthedocs.io/en/latest/services.html#servicethredds

>  The default methodology employed to categorize these two types of content is represented by the below configuration.

**Comment by MatProv** [comment]
run tests

File: birdhouse/config/magpie/providers.cfg.template
Comment (tlvu) [inline birdhouse/config/magpie/providers.cfg.template:71]: Looks good now but you sure `.ncml` do not need to be added to this list, as I mentioned here https://github.com/bird-house/birdhouse-deploy/pull/107#discussion_r530617162?  Your call, you know this best.

**Review by tlvu — APPROVED** [review]
LGTM but look at my comment.

File: birdhouse/config/magpie/providers.cfg.template
Comment (fmigneault) [reply birdhouse/config/magpie/providers.cfg.template:71]: Depends what we want to achieve. 
If we want to create 2 "resources", one of `.nc` and another for `.ncml`, it should be added. 
But if we want to have the same "permission set" for both these extension as if they were the same file, then this regex does just that.

File: birdhouse/config/magpie/providers.cfg.template
Comment (fmigneault) [reply birdhouse/config/magpie/providers.cfg.template:71]: The general idea of that regex really is to map file/path to a same resource. 

So regex ` ".+\\.nc"` will match both `.nc` and `.ncml`. 
This means that browse/read/write permissions will apply to the same resource for both paths: 
`[...]/somedir/somefile.nc`
`[...]/somedir/somefile.ncml`

Without the regex, those two file would have 2 distinct Magpie resources, meaning that you would need to duplicate the permissions set each time to replicate access control over the same file for different "view formats".

When there are no regex matched, the filename are looked for literally, so the `.txt` don't need a regex.
A specific resource must be created if you want to set a premission specifically on it, but must of the time I think the parent directory with "recursive" will be used for that case (license file to its contents)?

Comment (tlvu) [reply birdhouse/config/magpie/providers.cfg.template:71]: > So regex ` ".+\\.nc"` will match both `.nc` and `.ncml`.
> This means that browse/read/write permissions will apply to the same resource for both paths:
> `[...]/somedir/somefile.nc`
> `[...]/somedir/somefile.ncml`

For us the `.nc` and `.ncml` files are different file so different resource?  So new regex?  Here for example a `.ncml` file https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/catalog/datasets/gridded_obs/catalog.html?dataset=datasets/gridded_obs/nrcan_v2.ncml  So if you think it needs a new regex, please add it.

Comment (fmigneault) [reply birdhouse/config/magpie/providers.cfg.template:72]: In that case, yes it should be added to avoid truncating the extension to `.nc`. 
Added in f448778

File: birdhouse/config/magpie/providers.cfg.template
Comment (fmigneault) [reply birdhouse/config/magpie/providers.cfg.template:72]: done 
f448778

**Review by tlvu — APPROVED** [review]
LGTM, merge when ready.

**Comment by tlvu** [comment]
@fmigneault I triggered https://daccs-jenkins.crim.ca/job/DACCS-iac/121/ on the branch in this PR because I wanted to test the new auth notebook but then I noticed the standard test suite is broken as well (https://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/master/119/console)

`catalog_search.ipynb` failed which indicates there might be problem with the previous crawling operation.

`pavics_thredds.ipynb` failed with `HTTPError: 401 Client Error: Unauthorized for url: https://host-140-216.rdext.crim.ca/magpie/signin` for `auth = MagpieAuth("https://host-140-216.rdext.crim.ca/magpie", "authtest", "authtest1234")` so looks like user has not been created properly?

Logs in DACCS-iac for user creation `08-Dec-20 17:04:59 - INFO - Output results sent to [/tmp/create_magpie_users/magpie_create_users_log__20201208__170459.txt]` is pretty useless.  Is it possible to add an option to echo this output to STDOUT instead of keeping it inside the log file like that?

Ping @MatProv in case known error with infra.

**Comment by tlvu** [comment]
Oh by the way @MatProv the last commit https://github.com/bird-house/birdhouse-deploy/pull/107/commits/f448778695a07a076d044a84497a1887ca63f3e0 in this PR by Francis did not triggered the pipeline, is that normal?  The previous ones did.

**Comment by MatProv** [comment]
> Oh by the way @MatProv the last commit [f448778](https://github.com/bird-house/birdhouse-deploy/commit/f448778695a07a076d044a84497a1887ca63f3e0) in this PR by Francis did not triggered the pipeline, is that normal? The previous ones did.

@tlvu Should now be fixed. It seems to be be caused by the fact that crim-jenkins-bot wasn't a member of crim-ca org. Even if crim-ca members are whitelisted for the auto PR creation, the bot wasn't able to know if, for example, Francis was a member (private information).

@fmigneault Next time you push a commit or open a PR we will see if problem is gone.

**Comment by tlvu** [comment]
> Should now be fixed. It seems to be be caused by the fact that crim-jenkins-bot wasn't a member of crim-ca org. Even if crim-ca members are whitelisted for the auto PR creation, the bot wasn't able to know if, for example, Francis was a member (private information).

That's odd, the 2 previous pushes from Francis in this  PR did triggered the pipeline.  Look at https://github.com/bird-house/birdhouse-deploy/pull/107/commits/6cbfe850d2997d994318eb9795149d19e5f3b3b8 and https://github.com/bird-house/birdhouse-deploy/pull/107/commits/6e9b1d8764baf386197428f599e4e325d79226e1.

**Comment by MatProv** [comment]
@tlvu 
> `pavics_thredds.ipynb` failed with `HTTPError: 401 Client Error: Unauthorized for url: https://host-140-216.rdext.crim.ca/magpie/signin` for `auth = MagpieAuth("https://host-140-216.rdext.crim.ca/magpie", "authtest", "authtest1234")` so looks like user has not been created properly?

Looking at the file, authtest user is strangely already created:

`ubuntu@daccs-instance-24483-daccsci:~$ cat /tmp/create_magpie_users/magpie_create_users_log__20201208__170459.txt

USERNAME     PASSWORD         RESULT                                                     
___________________________________________________________________________________________

authtest     authtest1234     409 : User name matches an already existing user name. `

> Logs in DACCS-iac for user creation `08-Dec-20 17:04:59 - INFO - Output results sent to [/tmp/create_magpie_users/magpie_create_users_log__20201208__170459.txt]` is pretty useless. Is it possible to add an option to echo this output to STDOUT instead of keeping it inside the log file like that?

This echo to STDOUT would be located in `scripts/create-magpie-authtest-user`, right after authtest user creation.

> Ping @MatProv in case known error with infra.

I haven't been able to reproduce on `master` branch. DACCS-iac builds 122 and 123 passed with no infra/magpie permission issue, so this 401 would occur only in `magpie-thredds`.

**Comment by MatProv** [comment]
> That's odd, the 2 previous pushes from Francis in this PR did triggered the pipeline. 

@tlvu Actually both of these builds were triggered by a `run tests` comment (that I removed, after build was triggered), that's why.

**Comment by tlvu** [comment]
> Looking at the file, authtest user is strangely already created (cat of the txt file):
> 
> `409 : User name matches an already existing user name. `
>

This makes no sense, the user could not have already exist if this is a fresh new instance.  @MatProv could host host-140-216.rdext.crim.ca have been a recycled host and not fresh?
 
> This echo to STDOUT would be located in `scripts/create-magpie-authtest-user`, right after authtest user creation.
> 

So if a new `magpie_cli` option `--log-to-stdout` existed, it would be used in `scripts/create-magpie-authtest-user` but currently there is no such options.

> I haven't been able to reproduce on `master` branch. DACCS-iac builds 122 and 123 passed with no infra/magpie permission issue, so this 401 would occur only in `magpie-thredds`.

So something fishy with this new version of Magpie then.  @fmigneault next time with a new PR, please ensure the DACCS-iac pipeline passes.

**Comment by tlvu** [comment]
> > That's odd, the 2 previous pushes from Francis in this PR did triggered the pipeline.
> 
> @tlvu Actually both of these builds were triggered by a `run tests` comment (that I removed, after build was triggered), that's why.

@MatProv OH !  Leaving `run tests` behind would have been much less confusing.  I think editing history in general is bad.  Same can be said for real life history.

**Comment by MatProv** [comment]
> Leaving `run tests` behind would have been much less confusing. 

@tlvu Yes exactly, I did this to avoid spamming but it is of course harder to know that it was the comment that triggered the build.. when there is no more comment! I'll leave `run tests` comment there next time.

**Comment by fmigneault** [comment]
@tlvu @MatProv 
There is nothing new with Magpie user creation for a very long time. Only thing that changed was the password length which was already resolved by Mathieu. 
Since the response is a conflict, the user does exist, and is not a invalid handling on Magpie's part. 
Already existing user could be caused by timing between multiple tests that use the same user? If they all assume they create the user "from scratch" but it is repeated in many places without cleanup, it will exist in the db.

**Comment by fmigneault** [comment]
Tests execution details: 
https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/55#issuecomment-775493097

**Comment by MatProv** [comment]
run tests

**Comment by MatProv** [comment]
run tests

**Comment by tlvu** [comment]
@fmigneault FYI merge with `master` again to get this PR https://github.com/bird-house/birdhouse-deploy/pull/132  Jenkins passing (https://daccs-jenkins.crim.ca/job/DACCS-iac/237/).  Jenkins was broken during my vacation.

**Comment by fmigneault** [comment]
@tlvu Still getting the same errors caused by hostname diff of notebook results

snippet below, 
full log: http://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/magpie-func-tests/28/execution/node/28/log/
running: https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/55
with latest changes from: https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/63

```
[...]
10:15:23         "  <wps:Process wps:processVersion=\"4.4.1.1\">\n",
10:15:23         "    <ows:Identifier>ncdump</ows:Identifier>\n",
10:15:23         "    <ows:Title>NCDump</ows:Title>\n",
10:15:23  @@ -195,7 +195,7 @@
10:15:23         "      <ows:Identifier>output</ows:Identifier>\n",
10:15:23         "      <ows:Title>NetCDF Metadata</ows:Title>\n",
10:15:23         "      <ows:Abstract>NetCDF Metadata</ows:Abstract>\n",
10:15:23  -      "      <wps:Reference xlink:href=\"https://pavics.ouranos.ca:443/wpsoutputs/hummingbird/1df79124-3173-11e9-b38b-0242ac19000f/nc_dump_zKu8Os.txt\" mimeType=\"text/plain\"/>\n",
10:15:23  +      "      <wps:Reference xlink:href=\"https://host-140-29.rdext.crim.ca:443/wpsoutputs/hummingbird/1df79124-3173-11e9-b38b-0242ac19000f/nc_dump_zKu8Os.txt\" mimeType=\"text/plain\"/>\n",
10:15:23         "    </wps:Output>\n",
10:15:23         "  </wps:ProcessOutputs>\n",
10:15:23         "</wps:ExecuteResponse>\n"
10:15:23  @@ -217,7 +217,7 @@
10:15:23        "output_type": "stream",
10:15:23        "text": [
10:15:23         "Status:  ProcessSucceeded\n",
10:15:23  -      "https://pavics.ouranos.ca:443/wpsoutputs/hummingbird/1df79124-3173-11e9-b38b-0242ac19000f.xml\n"
10:15:23  +      "https://host-140-29.rdext.crim.ca:443/wpsoutputs/hummingbird/1df79124-3173-11e9-b38b-0242ac19000f.xml\n"
10:15:23        ]
10:15:23       }
10:15:23      ],
10:15:23  @@ -237,7 +237,7 @@
10:15:23        "output_type": "stream",
10:15:23        "text": [
10:15:23         "Output reference :\n",
10:15:23  -      "* https://pavics.ouranos.ca:443/wpsoutputs/hummingbird/c53de772-33ca-11e9-9a7a-0242ac19000f/nc_dump_RNoL6I.txt\n",
10:15:23  +      "* https://host-140-29.rdext.crim.ca:443/wpsoutputs/hummingbird/c53de772-33ca-11e9-9a7a-0242ac19000f/nc_dump_RNoL6I.txt\n",
10:15:23         "\n",
10:15:23         "NCDUMP results :\n",
[...]
```

**Comment by tlvu** [comment]
> ```
> [...]
> 10:15:23         "  <wps:Process wps:processVersion=\"4.4.1.1\">\n",
> 10:15:23         "    <ows:Identifier>ncdump</ows:Identifier>\n",
> 10:15:23         "    <ows:Title>NCDump</ows:Title>\n",
> 10:15:23  @@ -195,7 +195,7 @@
> 10:15:23         "      <ows:Identifier>output</ows:Identifier>\n",
> 10:15:23         "      <ows:Title>NetCDF Metadata</ows:Title>\n",
> 10:15:23         "      <ows:Abstract>NetCDF Metadata</ows:Abstract>\n",
> 10:15:23  -      "      <wps:Reference xlink:href=\"https://pavics.ouranos.ca:443/wpsoutputs/hummingbird/1df79124-3173-11e9-b38b-0242ac19000f/nc_dump_zKu8Os.txt\" mimeType=\"text/plain\"/>\n",
> 10:15:23  +      "      <wps:Reference xlink:href=\"https://host-140-29.rdext.crim.ca:443/wpsoutputs/hummingbird/1df79124-3173-11e9-b38b-0242ac19000f/nc_dump_zKu8Os.txt\" mimeType=\"text/plain\"/>\n",
> 10:15:23         "    </wps:Output>\n",
> 10:15:23         "  </wps:ProcessOutputs>\n",
> 10:15:23         "</wps:ExecuteResponse>\n"
> 10:15:23  @@ -217,7 +217,7 @@
> 10:15:23        "output_type": "stream",
> 10:15:23        "text": [
> 10:15:23         "Status:  ProcessSucceeded\n",
> 10:15:23  -      "https://pavics.ouranos.ca:443/wpsoutputs/hummingbird/1df79124-3173-11e9-b38b-0242ac19000f.xml\n"
> 10:15:23  +      "https://host-140-29.rdext.crim.ca:443/wpsoutputs/hummingbird/1df79124-3173-11e9-b38b-0242ac19000f.xml\n"
> 10:15:23        ]
> 10:15:23       }
> 10:15:23      ],
> 10:15:23  @@ -237,7 +237,7 @@
> 10:15:23        "output_type": "stream",
> 10:15:23        "text": [
> 10:15:23         "Output reference :\n",
> 10:15:23  -      "* https://pavics.ouranos.ca:443/wpsoutputs/hummingbird/c53de772-33ca-11e9-9a7a-0242ac19000f/nc_dump_RNoL6I.txt\n",
> 10:15:23  +      "* https://host-140-29.rdext.crim.ca:443/wpsoutputs/hummingbird/c53de772-33ca-11e9-9a7a-0242ac19000f/nc_dump_RNoL6I.txt\n",
> 10:15:23         "\n",
> 10:15:23         "NCDUMP results :\n",
> [...]
> ```

This is simply the output of `git diff` showing the replacement of `pavics.ouranos.ca` to your host `host-140-29.rdext.crim.ca` worked.  It's not counted as error at all.

> 
> Also got problems caused by unique test ids like follows:
> 
> ```
> 10:23:10   assert reference_output == test_output failed:
> 10:23:10  
> 10:23:10    'Create test ... user... OK\n' == 'Create test ...reation: OK\n'
> 10:23:10    - Create test group and test user... 
> 10:23:10    + Create test group and test user... OK
> 10:23:10    ?                                    ++
> 10:23:10    -   Will use TEST_USER_NAME:  [test-user-e3bba7b5-aa17-49de-912d-1f2b0b78f166]
> 10:23:10    -   Will use TEST_GROUP_NAME: [test-auth-65bc6430-220a-4dc3-97b8-5582f46fea28]
> 10:23:10    - Creation: OK
> ```

In this case, you can just add `# NBVAL_IGNORE_OUTPUT` to the first line of the cell that produce the output and it won't be flagged as error anymore.

Same for other cells that produce random hashes.   Either that or you can add new regexps to the file https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/blob/71ed13be63a4070dc70be4333d370cf678b54a98/notebooks/output-sanitize.cfg to escape those random hashes.

**Comment by tlvu** [comment]
From https://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/magpie-func-tests/28/console

I also see this `test_thredds.ipynb` failure, maybe you have to refresh the output of your notebook to match the code?

```
10:23:10  __________________ notebooks-auth/test_thredds.ipynb::Cell 7 ___________________

(...)

10:23:10  Traceback:
10:23:10  Missing output fields from running code: {'evalue', 'ename'}
```

`catalog_search.ipynb` also failed with a not helpful error `ProcessFailed: Sorry, process failed.`, check whether the crawling worked.  The crawling is triggered here https://github.com/bird-house/birdhouse-deploy/blob/091bf47d4ffac61f9c6c5aaee35fe551acc7848f/birdhouse/scripts/trigger-pavicscrawler#L31.  Crawler is triggered before the start of the Jenkins testsuite.  @MatProv knows the full sequence in the automated pipefline and will be able to assist you.


```
10:23:10  _____ pavics-sdi-master/docs/source/notebooks/catalog_search.ipynb::Cell 1 _____
10:23:10  Notebook cell execution failed
10:23:10  Cell 1: Cell execution caused an exception
10:23:10  
10:23:10  Input:
10:23:10  # NBVAL_IGNORE_OUTPUT
10:23:10  
10:23:10  resp = wps.pavicsearch(constraints="variable:tasmax,project:CMIP5,experiment:rcp45,model:MPI-ESM-MR,institute:MPI-M,frequency:mon", limit=100, type="File")
10:23:10  [result, files] = resp.get(asobj=True)
10:23:10  files
10:23:10  
10:23:10  Traceback:
10:23:10  
10:23:10  ---------------------------------------------------------------------------
10:23:10  ProcessFailed                             Traceback (most recent call last)
10:23:10  <ipython-input-2-327d69568721> in <module>
10:23:10        2 
10:23:10        3 resp = wps.pavicsearch(constraints="variable:tasmax,project:CMIP5,experiment:rcp45,model:MPI-ESM-MR,institute:MPI-M,frequency:mon", limit=100, type="File")
10:23:10  ----> 4 [result, files] = resp.get(asobj=True)
10:23:10        5 files
10:23:10  
10:23:10  /opt/conda/envs/birdy/lib/python3.7/site-packages/birdy/client/outputs.py in get(self, asobj)
10:23:10       28         if not self.isSucceded():
10:23:10       29             # TODO: add reason for failure
10:23:10  ---> 30             raise ProcessFailed("Sorry, process failed.")
10:23:10       31         return self._make_output(asobj)
10:23:10       32 
10:23:10  
10:23:10  ProcessFailed: Sorry, process failed.
```

For `pavics_thredds.ipynb` failure to access secured content with user `authest/authtest1234`, is that test user created with the appropriate password?  That user creation is here https://github.com/bird-house/birdhouse-deploy/blob/091bf47d4ffac61f9c6c5aaee35fe551acc7848f/birdhouse/scripts/create-magpie-authtest-user#L13-L23.  That user creation step is also done before the start of the Jenkins test suite.  Again @MatProv will be able to assist you if there are issues with that user creation during the automated pipeline.

```
10:23:10  _____ pavics-sdi-master/docs/source/notebooks/pavics_thredds.ipynb::Cell 2 _____
10:23:10  Notebook cell execution failed
10:23:10  Cell 2: Cell execution caused an exception
10:23:10  
10:23:10  Input:
10:23:10  import requests
10:23:10  from requests_magpie import MagpieAuth
10:23:10  
10:23:10  secured_url = "https://host-140-29.rdext.crim.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/testdata/secure/tasmax_Amon_MPI-ESM-MR_rcp45_r2i1p1_200601-200612.nc"
10:23:10  auth = MagpieAuth("https://host-140-29.rdext.crim.ca/magpie", "authtest", "authtest1234")
10:23:10  
10:23:10  # Open session
10:23:10  session = requests.Session()
10:23:10  session.auth = auth
10:23:10  
10:23:10  # Open a Pydap data store and pass it to xarray
10:23:10  store = xr.backends.PydapDataStore.open(secured_url, session=session)
10:23:10  ds = xr.open_dataset(store, decode_cf=False) # Attributes are problematic with this file. 
10:23:10  ds
10:23:10  
10:23:10  Traceback:
10:23:10  
10:23:10  ---------------------------------------------------------------------------
10:23:10  HTTPError                                 Traceback (most recent call last)
10:23:10  <ipython-input-3-b983054e0439> in <module>
10:23:10       10 
10:23:10       11 # Open a Pydap data store and pass it to xarray
10:23:10  ---> 12 store = xr.backends.PydapDataStore.open(secured_url, session=session)
10:23:10       13 ds = xr.open_dataset(store, decode_cf=False) # Attributes are problematic with this file.
10:23:10       14 ds
10:23:10  
10:23:10  /opt/conda/envs/birdy/lib/python3.7/site-packages/xarray/backends/pydap_.py in open(cls, url, session)
10:23:10       76         import pydap.client
10:23:10       77 
10:23:10  ---> 78         ds = pydap.client.open_url(url, session=session)
10:23:10       79         return cls(ds)
10:23:10       80 
10:23:10  
10:23:10  /opt/conda/envs/birdy/lib/python3.7/site-packages/pydap/client.py in open_url(url, application, session, output_grid, timeout)
10:23:10       65     """
10:23:10       66     dataset = DAPHandler(url, application, session, output_grid,
10:23:10  ---> 67                          timeout).dataset
10:23:10       68 
10:23:10       69     # attach server-side functions
10:23:10  
10:23:10  /opt/conda/envs/birdy/lib/python3.7/site-packages/pydap/handlers/dap.py in __init__(self, url, application, session, output_grid, timeout)
10:23:10       52         ddsurl = urlunsplit((scheme, netloc, path + '.dds', query, fragment))
10:23:10       53         r = GET(ddsurl, application, session, timeout=timeout)
10:23:10  ---> 54         raise_for_status(r)
10:23:10       55         if not r.charset:
10:23:10       56             r.charset = 'ascii'
10:23:10  
10:23:10  /opt/conda/envs/birdy/lib/python3.7/site-packages/pydap/net.py in raise_for_status(response)
10:23:10       37             detail=response.status+'\n'+response.text,
10:23:10       38             headers=response.headers,
10:23:10  ---> 39             comment=response.body
10:23:10       40         )
10:23:10       41 
10:23:10  
10:23:10  HTTPError: 401 Unauthorized
10:23:10  <?xml version="1.0" encoding="utf-8"?>
10:23:10  <ExceptionReport version="1.0.0"
10:23:10      xmlns="http://www.opengis.net/ows/1.1"
10:23:10      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
10:23:10      xsi:schemaLocation="http://www.opengis.net/ows/1.1 http://schemas.opengis.net/ows/1.1.0/owsExceptionReport.xsd">
10:23:10      <Exception exceptionCode="NoApplicableCode" locator="AccessForbidden">
10:23:10          <ExceptionText>Not authorized to access this resource. User does not meet required permissions.</ExceptionText>
10:23:10      </Exception>
10:23:10  </ExceptionReport>
```

**Comment by tlvu** [comment]
For the following `pavics_thredds.ipynb` error, you simply have to refresh the notebook output, see the comment `# This should fail but doesn't at the moment.` in the code.

```
10:23:10  _____ pavics-sdi-master/docs/source/notebooks/pavics_thredds.ipynb::Cell 1 _____
10:23:10  Notebook cell execution failed
10:23:10  Cell 1: Cell execution caused an exception
10:23:10  
10:23:10  Input:
10:23:10  secured_url = "https://host-140-29.rdext.crim.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/testdata/secure/tasmax_Amon_MPI-ESM-MR_rcp45_r2i1p1_200601-200612.nc"
10:23:10  
10:23:10  # This should fail but doesn't at the moment. 
10:23:10  ds = xr.open_dataset(secured_url, decode_cf=False)
10:23:10  
10:23:10  Traceback:
10:23:10  
10:23:10  ---------------------------------------------------------------------------
10:23:10  KeyError                                  Traceback (most recent call last)
10:23:10  /opt/conda/envs/birdy/lib/python3.7/site-packages/xarray/backends/file_manager.py in _acquire_with_cache_info(self, needs_lock)
10:23:10      198             try:
10:23:10  --> 199                 file = self._cache[self._key]
10:23:10      200             except KeyError:
10:23:10  
10:23:10  /opt/conda/envs/birdy/lib/python3.7/site-packages/xarray/backends/lru_cache.py in __getitem__(self, key)
10:23:10       52         with self._lock:
10:23:10  ---> 53             value = self._cache[key]
10:23:10       54             self._cache.move_to_end(key)
10:23:10  
10:23:10  KeyError: [<class 'netCDF4._netCDF4.Dataset'>, ('https://host-140-29.rdext.crim.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/testdata/secure/tasmax_Amon_MPI-ESM-MR_rcp45_r2i1p1_200601-200612.nc',), 'r', (('clobber', True), ('diskless', False), ('format', 'NETCDF4'), ('persist', False))]
10:23:10  
(...)
10:23:10  
10:23:10  OSError: [Errno -78] NetCDF: Authorization failure: b'https://host-140-29.rdext.crim.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/testdata/secure/tasmax_Amon_MPI-ESM-MR_rcp45_r2i1p1_200601-200612.nc'
```

**Comment by tlvu** [comment]
I think we should just merge this branch (magpie-thredds) into the base branch (bump_magpie_and_twitcher_to_3.2.0) so we only have 1 PR that upgrade Magpie, it's less confusing and make the review process simpler.  So merge when you are ready, I approve :D

**Comment by fmigneault** [comment]
> I think we should just merge this branch (magpie-thredds) into the base branch (bump_magpie_and_twitcher_to_3.2.0) so we only have 1 PR that upgrade Magpie, it's less confusing and make the review process simpler. So merge when you are ready, I approve :D

I changed the base so that we only use that PR. The one will be included automatically.

**Comment by fmigneault** [comment]
@dbyrns @tlvu @MatProv 
We got a winner 🎉 
http://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/magpie-func-tests/38/execution/node/28/log/
using Ouranosinc/PAVICS-e2e-workflow-tests#55 and Ouranosinc/pavics-sdi#209

I see failing tests for raven, but I'm not sure if this is normal or not as even our master branch doesn't evaluate them at the moment.

**Comment by Zeitsperre** [comment]
@fmigneault `raven` master was updated a few minutes ago. The new notebooks in that repo require a tagged version to work (numerous updates to dependencies). This should be coming soon.

**Comment by huard** [comment]
Don't wait on Raven notebooks to pass to merge this PR.

**Comment by tlvu** [comment]
@fmigneault In fact, you pull the trigger too fast man.  We have never done any upgrade testing !  All the Jenkins pipeline done was test a freshly new deployment.

**Comment by fmigneault** [comment]
> @fmigneault In fact, you pull the trigger too fast man. We have never done any upgrade testing ! All the Jenkins pipeline done was test a freshly new deployment.

https://hirondelle.crim.ca/twitcher/ows/proxy/thredds/
working as intended 

AFAIK this is the closest that can be tested before deploy to prod

**Comment by tlvu** [comment]
> > @fmigneault In fact, you pull the trigger too fast man. We have never done any upgrade testing ! All the Jenkins pipeline done was test a freshly new deployment.
> 
> https://hirondelle.crim.ca/twitcher/ows/proxy/thredds/
> working as intended
> 
> AFAIK this is the closest that can be tested before deploy to prod

Since we operate the production deployment, it would be kind if you let us do the merge this in repo so we have the chance to test against our datasets.  Remember all the DB migration problems on an existing instance with the previous magpie upgrade?

**Comment by tlvu** [comment]
@nikola-rados FYI, if you want to block the autodeploy on your PAVICS instance, you just have to create a new uncommitted file in your birdhouse-deploy checkout and the autodeploy will refuse to deploy new changes.  Hope you still have the time to do this.

We have not yet tested the DB migration on existing instance of this PR.  All the testing was done on a fresh new instances.

**Comment by tlvu** [comment]
@fmigneault I propose a revert of this PR so you take your time to flush out all the upgrade randomness.

**Pull request merged**
